backend/kafka-queue/kafkaqueue.go

```diff
@@ -452,22 +453,37 @@ func (p *Queue) LogStats() {
 	}
 }
 
-func (p *Queue) serializeMessage(msg *Message) (compressed []byte, err error) {
+func (p *Queue) serializeMessage(msg RetryableMessage) (compressed []byte, err error) {
 	compressed, err = json.Marshal(&msg)
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to marshall json")
 	}
 	return
 }
 
-func (p *Queue) deserializeMessage(compressed []byte) (msg *Message, error error) {
+func (p *Queue) deserializeMessage(compressed []byte) (RetryableMessage, error) {
 	if int64(len(compressed)) >= p.MessageSizeBytes {
 		return nil, errors.New("message too large")
 	}
+	var msgType struct {
+		Type PayloadType
+	}
+	if err := json.Unmarshal(compressed, &msgType); err != nil {
```

Comment:
just curious, what's the performance impact of this extra unmarshal? is it as slow as the second unmarshal on line 482?

Reply:
there is probably some performance impact, could also unmarshal to the default message type and then only marshal to the new type in case it's a flattened message.

I think ideally this type information would be encoded in the key. Do we rely on the key for anything at the moment? Or can we manually choose a partition to use regardless of the key?